dask/array/core.py

```diff
@@ -3743,22 +3863,104 @@ def from_delayed(value, shape, dtype=None, meta=None, name=None):
     >>> array
-    dask.array<from-value, shape=(5,), dtype=float64, chunksize=(5,), chunktype=numpy.ndarray>
+    dask.array<from-delayed, shape=(5,), dtype=float64, chunksize=(5,), chunktype=numpy.ndarray>
     >>> array.compute()
     array([1., 1., 1., 1., 1.])
+
+    >>> orig = da.zeros((6, 4), chunks=2)
+    >>> delayeds = orig.to_delayed(optimize_graph=False)
+    >>> delayeds[0, 0] = dask.delayed(np.ones)((2, 2))
+    >>> delayeds[-1, 1] -= 1
+    >>> array = da.from_delayed(delayeds, like=orig)
+    >>> array.compute()
+    array([[ 1.,  1.,  0.,  0.],
+           [ 1.,  1.,  0.,  0.],
+           [ 0.,  0.,  0.,  0.],
+           [ 0.,  0.,  0.,  0.],
+           [ 0.,  0., -1., -1.],
+           [ 0.,  0., -1., -1.]])
+
+    See also
+    --------
+    Array.to_delayed
+    map_blocks
+    blockwise
     """
-    from dask.delayed import Delayed, delayed
-
-    if not isinstance(value, Delayed) and hasattr(value, "key"):
-        value = delayed(value)
-
-    name = name or "from-value-" + tokenize(value, shape, dtype, meta)
-    dsk = {(name,) + (0,) * len(shape): value.key}
-    chunks = tuple((d,) for d in shape)
-    # TODO: value._key may not be the name of the layer in value.dask
-    # This should be fixed after we build full expression graphs
-    graph = HighLevelGraph.from_collections(name, dsk, dependencies=[value])
-    return Array(graph, name, chunks, dtype=dtype, meta=meta)
+
+    if not isinstance(value, np.ndarray):
+        if shape is not None:
+            ndim = len(shape)
+        elif chunks is not None:
+            ndim = len(chunks)
+        elif like is not None:
+            ndim = like.ndim
+        else:
+            raise ValueError(
+                "One of `chunks`, `shape`, or `like` must be specified when giving a single delayed value."
+            )
+
+        value = np.array([value], dtype=object, ndmin=ndim)
+
+    if value.dtype.kind != "O":
+        raise TypeError(
+            "`from_delayed` expects a NumPy object array containing `dask.delayed` objects, "
+            f"not {value.dtype}"
+        )
+    if like is not None:
+        if not isinstance(like, Array):
+            raise TypeError(f"`like=` must be a dask Array, not {type(like)}")
+        if chunks is None:
+            chunks = like.chunks
+        if meta is None:
+            meta = like._meta
+
+    if chunks is None and shape is None:
+        if not allow_unknown_chunksizes:
+            raise ValueError(
+                "Chunksizes must be specified, either via `chunks=` or `like=`. "
+                "If the sizes of individual chunks are actually unkown, you can pass "
+                "`allow_unknown_chunksizes=True`, but some downstream operations may not work."
+            )
+        chunks = tuple((float("nan"),) * s for s in value.shape)
+    else:
+        chunks = normalize_chunks(
+            chunks, shape=shape, numblocks=value.shape, dtype=dtype
+        )
```

Comment:
I think part of what's making me nervous here and in #9416  is that this is a ton of work normalizing chunksizes and shapes, and it is *still* easy to get the wrong answer for anything but uniformly chunked, easily divisible arrays. So I can do everything correctly (for some definition of correct, i.e., pass a valid set of inputs), and still get an array with incorrect chunking:

```python
import dask.array as da
import numpy as np

from dask import delayed

arr = delayed(np.ones((10,10)))
arr2 = delayed(np.zeros((5,10)))

delayed_arr = np.array([[arr,], [arr2,]])  # stacked array has shape (15, 10)

darr = da.from_delayed(
    delayed_arr,
    meta = arr.dype,
    shape=(15,10),
)

darr.chunks  # ((8,7), (10,)) is wrong! It assumed even chunking given `shape`.
```

The above might not be fatal (maybe we just don't take `shape` here), but it does make me want to look closer at *exactly* what the chunking requirements are. I fear there might be no real option other than being extremely explicit (as in, the full tuple-of-tuples).

Reply:
Hm, that's a great point. Maybe the `shape` parameter is too easy to get wrong.

Without the shape parameter though, is it possible to get incorrect chunking as long as you pass an input that's correct?

For this example, passing `chunks=5`, `chunks=(10, 10)`, or `chunks=(5, 10)` would obviously be simply wrong. There's no way way to specify chunks correctly in this example without at least one tuple-of-tuples.

I don't mind _that_ much forcing the full tuple-of-tuples, but it is a bit annoying in simple cases where you know every chunk is 10x10, for example.

I could definitely see removing the `shape` parameter, but besides that, I think it's all just logic for a) making repeated tuples and b) validating that the chunks you gave matches up with the shape of the array you gave?

Reply:
> Without the shape parameter though, is it possible to get incorrect chunking as long as you pass an input that's correct?

Here is another example, which I believe is just a bug:

```python
import dask.array as da
import numpy as np

from dask import delayed


arr = da.ones((10, 10), chunks=3)  # Chunked with strips of length one on the edges
delayeds = arr.to_delayed()

darr = da.from_delayed(
    delayed_arr,
    like=arr
)
```
produces
```python-traceback
ValueError                                Traceback (most recent call last)
Input In [72], in <cell line: 10>()
      7 arr = da.ones((10, 10), chunks=3)  # Chunked with strips of length one on the edges
      8 delayeds = arr.to_delayed()
---> 10 darr = da.from_delayed(
     11     delayed_arr,
     12     like=arr
     13 )
     15 darr.chunks

File ~/dask/dask/dask/array/core.py:3926, in from_delayed(value, shape, chunks, dtype, meta, name, like, allow_unknown_chunksizes)
   3924     chunks = tuple((float("nan"),) * s for s in value.shape)
   3925 else:
-> 3926     chunks = normalize_chunks(
   3927         chunks, shape=shape, numblocks=value.shape, dtype=dtype
   3928     )
   3930 meta = meta_from_array(meta, ndim=len(chunks), dtype=dtype)
   3932 name = name or "from-delayed-" + tokenize(
   3933     value, chunks, meta, allow_unknown_chunksizes
   3934 )

File ~/dask/dask/dask/array/core.py:3141, in normalize_chunks(chunks, shape, numblocks, limit, dtype, previous_chunks)
   3139 if numblocks is not None:
   3140     if not all(len(c) == nb for c, nb in zip(chunks, numblocks)):
-> 3141         raise ValueError(
   3142             f"Chunks do not match numblocks. Got {chunks=}, {numblocks=}"
   3143         )
   3145 return tuple(tuple(int(x) if not math.isnan(x) else x for x in c) for c in chunks)

ValueError: Chunks do not match numblocks
```

I suspect the above is fixable, but what I was aiming at there is that a user could pass `chunks=3` to `from_delayed()`, thinking "Aha! It has the same chunking, so I'll give the same argument!". But I would not necessarily expect `normalize_chunks` to be able to know that the last strip should have length one in that case. You could argue that it's user error, but to me it's quite-easy-to-hit user error.

There may be others, I haven't done any kind of exhaustiveness reasoning. But that is also what makes me skittish around adding another dimension to `normalize_chunks`.

> I don't mind _that_ much forcing the full tuple-of-tuples, but it is a bit annoying in simple cases where you know every chunk is 10x10, for example.

I think it might be unavoidable, to be honest. There is a fundamental difference here with other array-creation schemes in that dask has no flexibility in deciding chunking, rounding things one way or another, etc. I just don't think `normalize_chunks` was designed for this.